File: GroupAnagram/GroupAnagram.js


function GroupAnagram (strs){
   // Step 1: Create a HashMap<String, List<String>>
//         Key   → sorted version of the string
//         Value → list of original anagram strings

// Step 2: For each string in the input array:
//   - Convert the string to a char array
//   - Sort the char array
//   - Convert it back to a string → this becomes the key
//   - If the key exists in the map, add the string to its list
//   - Otherwise, create a new list and put it in the map

// Step 3: Repeat until all strings are processed

// Step 4: Return all the values of the map as a List<List<String>>
// “All anagrams produce the same sorted character sequence, so we use it as a HashMap key.”


    console.log("Groupanagram",strs)
}


GroupAnagram(["eat","tea","tan","ate","nat","bat"]);

File: GroupAnagram/GroupAnagram.java
package GroupAnagram;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroupAnagram {

    List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for (String s : strs) {
            int[] count = new int[26];

            for (char c : s.toCharArray()) {
                count[c - 'a']++;
            }

            StringBuilder sb = new StringBuilder();
            for (int i : count) {
                sb.append('#').append(i);
            }

            String key = sb.toString();

            if(!map.containsKey(key)){
               map.put(key, new ArrayList<String>());
            }
            map.get(key).add(s);
            // create list if not present
            // map.putIfAbsent(key, new ArrayList<>());
            // map.get(key).add(s);
        }

        return new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        GroupAnagram g1 = new GroupAnagram();
        List<List<String>> value =
                g1.groupAnagrams(new String[]{"eat", "tea", "tan", "ate", "nat", "bat"});
        System.out.println(value);
    }
}
